xml-conduit/test/unit.hs

```diff
@@ -736,6 +740,37 @@ resolvedInline = do
     Res.Document _ root2 _ <- return $ Res.parseLBS_ Res.def "<!DOCTYPE foo [<!ENTITY bar \"baz\">]><foo bar='&bar;'/>"
     root2 @?= Res.Element "foo" (Map.singleton "bar" "baz") []
 
+resolvedInlineRecursive :: Assertion
+resolvedInlineRecursive = do
+    Res.Document _ root _ <- return $ Res.parseLBS_ Res.def
+      "<!DOCTYPE foo [<!ENTITY bim \"baz\"><!ENTITY bar \"&bim;&#73;&amp;\">]><foo>&bar;</foo>"
+    root @?= Res.Element "foo" Map.empty [Res.NodeContent "bazI&"]
+
+resolvedInlineLoop :: Assertion
+resolvedInlineLoop = do
+    res <- return $ Res.parseLBS Res.def
+           "<!DOCTYPE foo [<!ENTITY bim \"&bim;\">]><foo>&bim;</foo>"
+    Left (toException $ Res.UnresolvedEntityException (Set.fromList ["bim"]))
```

Comment:
[Optional] We may want to provide more info in the exception structure, to distinguish between:
- undefined entities
- recursive entities
- entities for which max expansion size has been reached

Reply:
Currently we don't raise any exception in `resolveEntities` itself (indeed, it's just a pure function and we can't raise the exception directly there). Instead, when an entity cannot be successfully resolved, either because it's not defined, or because it loops, or because its expansion is too large, we just pass through a `ContentEntity` token.  The exception gets raised at a later stage when this token is encountered (by some code I haven't touched, I guess `fromEvents` and `fromXMLElement` in Text.XML.).

To allow that code to produce different exceptions for these different cases, we'd need to have an additional parameter on `ContentEntity` (if that's just an internal type then maybe that's okay).

Maybe something like

```
ContentEntity EntityStatus Text
```
where
```
data EntityStatus = Unresolved | Loop | ExceedsMaxSize
```

Reply:
If you want to go this route, you'd need to define for me what exceptions you want to produce from `fromEvents` for each of these cases.

Reply:
Indeed, it seems it would require many more changes, that are beyond the scope of this pull-request. I'll have a look after merging it.